components/usage/pkg/apiv1/billing.go

```diff
@@ -165,6 +167,76 @@ func (s *BillingService) CreateStripeCustomer(ctx context.Context, req *v1.Creat
 	}, nil
 }
 
+func (s *BillingService) CreateStripeSubscription(ctx context.Context, req *v1.CreateStripeSubscriptionRequest) (*v1.CreateStripeSubscriptionResponse, error) {
+	attributionID, err := db.ParseAttributionID(req.GetAttributionId())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid attribution ID %s", attributionID)
+	}
+
+	customer, err := s.GetStripeCustomer(ctx, &v1.GetStripeCustomerRequest{Identifier: &v1.GetStripeCustomerRequest_AttributionId{AttributionId: string(attributionID)}})
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "Stripe customer with attribution ID %s not found", attributionID)
+	}
+
+	_, err = s.stripeClient.SetDefaultPaymentForCustomer(ctx, customer.Customer.Id, req.SetupIntentId)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Failed to set default payment for customer ID %s", customer.Customer.Id)
+	}
+
+	stripeCustomer, err := s.stripeClient.GetCustomer(ctx, customer.Customer.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	priceIdentifier := getPriceIdentifier(attributionID, stripeCustomer, s)
+	if priceIdentifier == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid currency %s for customer ID %s", stripeCustomer.Metadata["preferredCurrency"], stripeCustomer.ID)
+	}
```

Comment:
Because here you only check the empty string, and return an error. You can increase cohesion of this handling by pushing that error into the `getPriceIdentifier`. It would look like this:

```suggestion
	priceID, err := getPriceIdentifier(attributionID, stripeCustomer, s)
	if err != nil {
	  // note we're not wrapping the error, as the `getPriceIdentifier` can do that for us, with more context on the actual failure.
	  return nil, err
	}
```

The `getPriceIdentifier` then becomes:
```golang

func getPriceIdentifier(attributionID db.AttributionID, stripeCustomer *stripe_api.Customer, s *BillingService) (string, error) {
	preferredCurrency := stripeCustomer.Metadata["preferredCurrency"]
	if stripeCustomer.Metadata["preferredCurrency"] == "" {
		log.
			WithField("stripe_customer_id", stripeCustomer.ID).
			Warn("No preferred currency set. Defaulting to USD")
	}

	entity, _ := attributionID.Values()

	switch entity {
		case db.AttributionEntity_User:
			switch preferredCurrency {
				case "EUR":
					return s.stripePrices.IndividualUsagePriceIDs.EUR, nil
				default:
					return s.stripePrices.IndividualUsagePriceIDs.USD, nil
			}

		case AttributionEntity_Team:
			switch preferredCurrency {
				case "EUR":
					return s.stripePrices.TeamUsagePriceIDs.EUR, nil
				default:
					return s.stripePrices.TeamUsagePriceIDs.USD, nil
			}

		default:
			return "", status.Errorf(codes.InvalidArgument, "Invalid currency %s for customer ID %s", stripeCustomer.Metadata["preferredCurrency"], stripeCustomer.ID)
	}
}

```

Reply:
Thanks this is much better, and good point about returning the error here instead.